File: src/main/java/com/brandonitaly/bedrockskins/client/StateManager.java
package com.brandonitaly.bedrockskins.client;

import com.google.gson.Gson;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.util.Collections;
import java.util.List;
import net.minecraft.client.Minecraft;

public final class StateManager {
    private StateManager() {}

    private static final Gson gson = new Gson();
    private static final File stateFile = new File(Minecraft.getInstance().gameDirectory, "bedrock_skins_state.json");

    public static BedrockSkinsState readState() {
        try {
            if (!stateFile.exists()) return new BedrockSkinsState(Collections.emptyList(), null);
            try (FileReader reader = new FileReader(stateFile)) {
                BedrockSkinsState st = gson.fromJson(reader, BedrockSkinsState.class);
                return st == null ? new BedrockSkinsState(Collections.emptyList(), null) : st;
            }
        } catch (Exception e) {
            e.printStackTrace();
            return new BedrockSkinsState(Collections.emptyList(), null);
        }
    }

    public static void saveState(List<String> favorites, String selected) {
        try {
            File tmp = new File(stateFile.getAbsolutePath() + ".tmp");
            try (FileWriter writer = new FileWriter(tmp)) {
                gson.toJson(new BedrockSkinsState(favorites, selected), writer);
            }
            if (tmp.exists()) {
                if (stateFile.exists()) stateFile.delete();
                tmp.renameTo(stateFile);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/com/brandonitaly/bedrockskins/client/StateManager.kt
package com.brandonitaly.bedrockskins.client

import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import net.minecraft.client.MinecraftClient
import java.io.File
import java.io.FileReader
import java.io.FileWriter

data class BedrockSkinsState(
    @SerializedName("favorites") val favorites: List<String> = emptyList(),
    @SerializedName("selected") val selected: String? = null
)

object StateManager {
    private val gson = Gson()
    private val stateFile: File = File(MinecraftClient.getInstance().runDirectory, "bedrock_skins_state.json")

    fun readState(): BedrockSkinsState {
        try {
            if (!stateFile.exists()) return BedrockSkinsState()
            FileReader(stateFile).use { reader ->
                return gson.fromJson(reader, BedrockSkinsState::class.java) ?: BedrockSkinsState()
            }
        } catch (e: Exception) {
            e.printStackTrace()
            return BedrockSkinsState()
        }
    }

    fun saveState(favorites: List<String>, selected: String?) {
        try {
            val tmp = File(stateFile.absolutePath + ".tmp")
            FileWriter(tmp).use { writer ->
                gson.toJson(BedrockSkinsState(favorites, selected), writer)
            }
            if (tmp.exists()) {
                if (stateFile.exists()) stateFile.delete()
                tmp.renameTo(stateFile)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}
